[KernelInfo] Implement new LLVM IR pass for GPU code analysis

jdenny-ornl opened the pull request:
This patch implements an LLVM IR pass, named kernel-info, that reports various statistics for codes compiled for GPUs.  The ultimate goal of these statistics to help identify bad code patterns and ways to mitigate them.  The pass operates at the LLVM IR level so that it can, in theory, support any LLVM-based compiler for programming languages supporting GPUs.  It has been tested so far with LLVM IR generated by Clang for OpenMP offload codes targeting NVIDIA GPUs and AMD GPUs.

By default, the pass runs at the end of LTO, and options like ``-Rpass=kernel-info`` enable its remarks.  Example `opt` and `clang` command lines appear in `llvm/docs/KernelInfo.rst`.  Remarks include summary statistics (e.g., total size of static allocas) and individual occurrences (e.g., source location of each alloca).  Examples of its output appear in tests in `llvm/test/Analysis/KernelInfo`.

jdenny-ornl committed:
[KernelInfo] Implement new LLVM IR pass for GPU code analysis

This patch implements an LLVM IR pass, named kernel-info, that reports
various statistics for codes compiled for GPUs.  The ultimate goal of
these statistics to help identify bad code patterns and ways to
mitigate them.  The pass operates at the LLVM IR level so that it can,
in theory, support any LLVM-based compiler for programming languages
supporting GPUs.  It has been tested so far with LLVM IR generated by
Clang for OpenMP offload codes targeting NVIDIA GPUs and AMD GPUs.

By default, the pass is disabled.  For convenience,
`-kernel-info-end-lto` inserts it at the end of LTO, and options like
`-Rpass=kernel-info` enable its remarks.  Example opt and clang
command lines appear in comments in
`llvm/include/llvm/Analysis/KernelInfo.h`.  Remarks include summary
statistics (e.g., total size of static allocas) and individual
occurrences (e.g., source location of each alloca).  Examples of its
output appear in tests in `llvm/test/Analysis/KernelInfo`.

tschuett commented:
Can you put an rst file to https://github.com/llvm/llvm-project/tree/main/llvm/docs instead of hiding everything in a header?

shiltian commented on llvm/test/Analysis/KernelInfo/openmp/amdgpu.ll:
```diff
@@ -0,0 +1,217 @@
+; See ./README.md for how to maintain the LLVM IR in this test.
```
The test case needs to be stripped.

jdenny-ornl replied:
What does stripped mean in this context?  Removed?  Reduced?

shiltian replied:
At least we don't need debug info?

jdoerfert replied:
We need debug info for the remarks location: `remark: test.c:15:9: `

jdenny-ornl replied:
Then we wouldn't be testing that remarks have correct source locations, right?

shiltian replied:
Oh I c. Didn't catch that part.

shiltian commented on llvm/include/llvm/Analysis/KernelInfo.h:
```diff
@@ -0,0 +1,148 @@
+  std::optional<int64_t> AmdgpuMaxNumWorkgroupsX;
+  std::optional<int64_t> AmdgpuMaxNumWorkgroupsY;
+  std::optional<int64_t> AmdgpuMaxNumWorkgroupsZ;
+  std::optional<int64_t> AmdgpuFlatWorkGroupSizeMin;
```
IIRC these are already part of AMDGPUAttributorPass?

jdoerfert replied:
They are propagated there, but also provided by the user.

jdenny-ornl replied:
I believe this comment was part of [the more general discussion here](https://github.com/llvm/llvm-project/pull/102944#issuecomment-2284621052).  Resolving.  Please reopen if necessary.

shiltian commented on llvm/include/llvm/Analysis/KernelInfo.h:
```diff
@@ -0,0 +1,148 @@
+
+  /// OpenMP Launch bounds.
+  ///@{
+  std::optional<int64_t> OmpTargetNumTeams;
```
Didn't we do that in OpenMPOpt?

jdenny-ornl replied:
I believe this comment was part of [the more general discussion here](https://github.com/llvm/llvm-project/pull/102944#issuecomment-2284621052).  Resolving.  Please reopen if necessary.

shiltian commented:
My general question is, the kernel info looks like highly "target" dependent. I'm not sure if it is a good idea to have them all combined together in a target "independent" manner.

jdoerfert commented on llvm/lib/Target/NVPTX/NVPTXTargetMachine.cpp:
```diff
@@ -238,6 +239,15 @@ void NVPTXTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
+
+  PB.registerFullLinkTimeOptimizationLastEPCallback(
+      [](ModulePassManager &PM, OptimizationLevel Level) {
+        if (KernelInfoEndLTO) {
```
Move the conditional outside.

jdenny-ornl replied:
Done.

jdoerfert commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -772,6 +773,15 @@ void AMDGPUTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
+
+  PB.registerFullLinkTimeOptimizationLastEPCallback(
+      [](ModulePassManager &PM, OptimizationLevel Level) {
+        if (KernelInfoEndLTO) {
```
Move the conditional outside.

jdenny-ornl replied:
Done.

shiltian commented:
An alternative is to have a base class with common info and then have target dependent part in sub classes.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,350 @@
+      }
+      remarkAlloca(ORE, F, *Alloca, StaticSize);
+    } else if (const CallBase *Call = dyn_cast<CallBase>(&I)) {
+      std::string CallKind;
+      std::string RemarkKind;
```
Small string

jdenny-ornl replied:
Done.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,350 @@
+      }
+      remarkCall(ORE, F, *Call, CallKind, RemarkKind);
+      if (const AnyMemIntrinsic *MI = dyn_cast<AnyMemIntrinsic>(Call)) {
+        if (MI->getDestAddressSpace() == 0) {
```
Would be more kosher to check TTI::getFlatAddressSpace

jdenny-ornl replied:
Done.  Should we change the remarks to say flat addrspace instead of addrspace(0)?

arsenm replied:
Yes

jdenny-ornl replied:
Done.

arsenm commented on llvm/test/Analysis/KernelInfo/addrspace0.ll:
```diff
@@ -0,0 +1,152 @@
+entry:
+  ; load
+  ; CHECK: remark: test.c:3:11: in function 'f', 'load' instruction accesses memory in addrspace(0)
+  %0 = load i32, ptr null, align 4, !dbg !6
```
Use named values in tests

jdenny-ornl replied:
Done.  But not there, which is specifically checking the case that the value is not named.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,350 @@
+
+static bool isKernelFunction(Function &F) {
+  // TODO: Is this general enough?  Consider languages beyond OpenMP.
+  return F.hasFnAttribute("kernel");
```
I don't think this attribute should even exist... Should be checking the set of calling conventions

jdenny-ornl replied:
Done.  I left the apparently OpenMP-specific stuff in OpenMPOpt.

jdenny-ornl replied:
Conversation continues at [this comment](https://github.com/llvm/llvm-project/pull/102944#discussion_r1906118317).

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,350 @@
+      parseFnAttrAsIntegerFields(F, "amdgpu-flat-work-group-size", 2);
+  KI.AmdgpuFlatWorkGroupSizeMin = AmdgpuFlatWorkGroupSize[0];
+  KI.AmdgpuFlatWorkGroupSizeMax = AmdgpuFlatWorkGroupSize[1];
+  auto AmdgpuWavesPerEu =
+      parseFnAttrAsIntegerFields(F, "amdgpu-waves-per-eu", 2);
```
Don't we already report information on this in the pass remarks in the AsmPrinter?

jdoerfert replied:
The AMDGPU remarks emit 0 for values of non-kernel functions, IIRC. Having a single location for all relevant remarks is also helpful. Only post-instruction selection & reg alloc stuff should be printed per-target (=for AMDGPU) late.

arsenm replied:
That's just a bug though

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,350 @@
+  // TODO: This would be more maintainable if there were an isGPU.
+  const std::string &TT = F.getParent()->getTargetTriple();
+  llvm::Triple T(TT);
+  if (!T.isAMDGPU() && !T.isNVPTX())
+    return KI;
```
Don't bother checking the triple?

jdenny-ornl replied:
You want it to trigger on cpu code as well?

arsenm replied:
It won't be added to the pass pipeline there, so it doesn't matter

jdenny-ornl replied:
You're right.  I forgot that case doesn't happen so easily in the current implementation.

jdenny-ornl replied:
Removed it.

jdoerfert commented:
> An alternative is to have a base class with common info and then have target dependent part in sub classes.

I don't see how that buys us much and I feel there is a cost to it (more files, more places, more boilerplate). What do we gain by splitting it up? In the existing code we can/should just print target dependent features (like waves-per-eu) only for AMDGPU and call it a day, no?

jdenny-ornl committed:
Move docs to KernelInfo.rst

jdenny-ornl committed:
Move conditional outside registration call

jdenny-ornl committed:
Use llvm::SmallString

jdenny-ornl committed:
Use TTI.getFlatAddressSpace for addrspace(0)

We have to be more careful about targets in the test suite now because
`getFlatAddressSpace` returns garbage for unsupported targets.

Should we change the remarks to say flat addrspace instead of
addrspace(0)?

jdenny-ornl committed:
Avoid repetition between amdgpu and nvptx tests

jdenny-ornl committed:
Use named values in tests

jdenny-ornl commented:
> Can you put an rst file to https://github.com/llvm/llvm-project/tree/main/llvm/docs instead of hiding everything in a header?

It's in `llvm/docs/KernelInfo.rst`.  Any suggestions on what other docs should link to it?

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,353 @@
+      }
+      remarkCall(ORE, F, *Call, CallKind, RemarkKind);
+      if (const AnyMemIntrinsic *MI = dyn_cast<AnyMemIntrinsic>(Call)) {
+        if (MI->getDestAddressSpace() == TTI.getFlatAddressSpace()) {
```
Cache the flat address space value in the class somewhere like InferAddressSpaces does instead of calling it each time

jdenny-ornl replied:
Done.

jdenny-ornl committed:
Say flat address space instead of addrspace(0)

jdenny-ornl committed:
Cache the flat address space

jdenny-ornl committed:
Link KernelInfo.rst from Passes.rst

jdenny-ornl committed:
Don't filter out cpus

-kernel-info-end-lto doesn't insert kernel-info for cpu modules.  If
the user explicitly specifies the pass for a cpu module, then it will
run now.

jdenny-ornl commented:
> > Can you put an rst file to https://github.com/llvm/llvm-project/tree/main/llvm/docs instead of hiding everything in a header?
> 
> It's in `llvm/docs/KernelInfo.rst`. Any suggestions on what other docs should link to it?

I linked it from Passes.rst.  I'm fine to move it if people think that's not the best place.

tschuett commented:
Short-term you could add an Offload.rst and link everything from there. Long-term you may ask for offload.llvm.org.

jdenny-ornl commented:
> Short-term you could add an Offload.rst and link everything from there. Long-term you may ask for offload.llvm.org.

Maybe, but it seems beyond the scope of this patch.  @jdoerfert, any opinion on the docs?

arsenm commented on llvm/include/llvm/Analysis/KernelInfo.h:
```diff
@@ -0,0 +1,121 @@
+#define LLVM_ANALYSIS_KERNELINFO_H
+
+#include "llvm/Analysis/OptimizationRemarkEmitter.h"
+#include "llvm/Analysis/TargetTransformInfo.h"
```
Shouldn't need these includes in the header

jdenny-ornl replied:
Done.

arsenm commented on llvm/include/llvm/Analysis/KernelInfo.h:
```diff
@@ -0,0 +1,121 @@
+class Function;
+
+/// Data structure holding function info for kernels.
+class KernelInfo {
```
Is there any real use of this beyond running the remark emitter? Sink this into the implementation file?

shiltian replied:
I'm curious as well. If there is no, I really don't think it should be in a target independent way that merges attributes from all targets together.

jdoerfert replied:
wrt. other uses: Likely, in the future. For now, we can sink it into the impl file.
wrt. merging attributes: There is really nothing to be won by splitting this into 3-4 files. It just adds complexity.

shiltian replied:
It doesn't have to be multiple files. The variadic argument pass has target dependent stuff in the same file as well. 

What I expect is a hierarchical modular data struct. Each target just minds their own business: add their own attributes in their own descriptor, implement their own analysis and print method, etc. For now it "seems" okay to have all of them in the same data structure and use if-else to print, because technically we only have two offloading targets, but this data structure will keep growing when we have more (looking at Intel), and for any of the target, most of the parts are useless.

For the future use, we can just have the analysis results at one place and return a reference.

jdenny-ornl replied:
> For now, we can sink it into the impl file.

In that case, does anyone see value in keeping the KernelInfoAnalysis class for now?  KernelInfoPrinter can just call KernelInfo directly.

jdenny-ornl replied:
> > For now, we can sink it into the impl file.
> 
> In that case, does anyone see value in keeping the KernelInfoAnalysis class for now? KernelInfoPrinter can just call KernelInfo directly.

I've dropped KernelInfoAnalysis for now.

jdenny-ornl replied:
@shiltian The [commit I just added](https://github.com/llvm/llvm-project/pull/102944/commits/62d494d9a9f13e5b58d71f083a6cb9f67f19579b) to address another concern probably also address your concern about target modularity.

jdenny-ornl replied:
Resolving this issue based on my previous comment here.  Please reopen if necessary.

arsenm commented on llvm/include/llvm/Analysis/KernelInfo.h:
```diff
@@ -0,0 +1,121 @@
+  static KernelInfo getKernelInfo(Function &F, FunctionAnalysisManager &FAM);
+
+  bool operator==(const KernelInfo &FPI) const {
+    return std::memcmp(this, &FPI, sizeof(KernelInfo)) == 0;
```
This is possibly wrong if there are any padding bytes anywhere

jdenny-ornl replied:
Sorry, those functions are unused.  Removed.  Thanks.

arsenm commented on llvm/include/llvm/Analysis/KernelInfo.h:
```diff
@@ -0,0 +1,121 @@
+
+  /// AMDGPU launch bounds.
+  ///@{
+  std::optional<int64_t> AmdgpuMaxNumWorkgroupsX;
+  std::optional<int64_t> AmdgpuMaxNumWorkgroupsY;
+  std::optional<int64_t> AmdgpuMaxNumWorkgroupsZ;
+  std::optional<int64_t> AmdgpuFlatWorkGroupSizeMin;
+  std::optional<int64_t> AmdgpuFlatWorkGroupSizeMax;
+  std::optional<int64_t> AmdgpuWavesPerEuMin;
+  std::optional<int64_t> AmdgpuWavesPerEuMax;
```
Instead of optional, just initialize to the maximum values?

jdenny-ornl replied:
How is that better?

arsenm replied:
Because these aren't optional and you avoid the noise of using optional. There is always a definitive range of values for these

jdenny-ornl replied:
> Because these aren't optional

How do you mean?  Sometimes they do not appear in the IR.

arsenm replied:
Just because there isn't an explicit attribute doesn't mean the value doesn't exist (actually the attributes can be used to specify impossible values, which then need to be logically clamped). The logical values always exist.

For reference, AMDGPUAttributor queries the logical values for these when doing the propagation: https://github.com/llvm/llvm-project/blob/f668708796b981733a5816f2efed0d5195af923d/llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp#L186

jdenny-ornl replied:
OK, so your argument isn't about std::optional vs. max value.  It's about reporting just what's explicit in the IR vs. the effective value.  Right?  @jdoerfert : What did you want to see here?

jdoerfert replied:
We should likely query the "default/logical" values when there is no explicit one. If we run after AMDGPUAttributor, we could ensure (or maybe it is already ensured) that we make all values explicit.

jdenny-ornl replied:
After merging a more recent main, I see KernelInfoPrinter run after AMDGPUAttributorPass when doing this:

```
$ clang -O3 -g -fopenmp --offload-arch=native test.c -foffload-lto \
    -Rpass=kernel-info -mllvm -kernel-info-end-lto \
    -Xoffload-linker --lto-debug-pass-manager
```

Only AmdgpuFlatWorkGroupSizeMin and AmdgpuFlatWorkGroupSizeMax are set.  Should I be seeing more?

jdenny-ornl replied:
I tried a real application (OpenMC) to get a larger sample of functions.  I found that only amdgpu-flat-work-group-size was set on some functions, only amdgpu-waves-per-eu was set on some functions, both were set on some functions, and amdgpu-max-num-workgroups was not set on any function.

Is this expected behavior for now?  If so, should AMDGPUAttributor be changed, or should I extend kernel-info to query the effective values?

arsenm replied:
I don't understand the context of these statements. Where did you see these? I don't what changed AMDGPUAttributor would need to make. amdgpu-flat-work-group-size and amdgpu-waves-per-eu are commonly set by users to some value.  amdgpu-max-num-workgroups  isn't really implemented.

After AMDGPUAttributor runs, the effective values will be propagated. What is the expected behavior if this pass runs before that? Is that what you were looking at?

jdenny-ornl replied:
As I mentioned above, I'm running KernelInfoPrinter after AMDGPUAttributorPass.

@jdoerfert said:

> We should likely query the "default/logical" values when there is no explicit one. If we run after AMDGPUAttributor, we could ensure (or maybe it is already ensured) that we make all values explicit.

Perhaps I misunderstood what the various occurrences of "we" mean.  I read it as, if AMDGPUAttributor were to make all values explicit, then KernelInfoPrinter wouldn't have to query the "default/logical" values.

jdenny-ornl replied:
Given that AMDGPUAttributor doesn't appear to always make these values explicit, I added [a commit](https://github.com/llvm/llvm-project/pull/102944/commits/62d494d9a9f13e5b58d71f083a6cb9f67f19579b) to query the values.

jdenny-ornl replied:
I think this issue is handled even though the implementation details are still under discussion elsewhere.  Resolving.  Please reopen if necessary.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,345 @@
+    }
+    R << "with ";
+    if (StaticSize)
+      R << "static size of " << itostr(StaticSize) << " bytes";
```
I think itostr is unnecessary, this should print int types as decimal anyway

jdenny-ornl replied:
It doesn't compile if I simply remove itostr.  Is that what you mean?

arsenm replied:
Yes, usually << just handles integers correctly

jdenny-ornl replied:
That doesn't seem to be true for OptimizationRemarkEmitter.  Please let me know if I've made a mistake somehow.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,345 @@
+      R << ", callee is";
+      StringRef Name = Callee->getName();
+      if (auto *SubProgram = Callee->getSubprogram()) {
+        if (SubProgram->isArtificial())
+          R << " artificial";
```
Is there a way to get a function name from debug info?

jdenny-ornl replied:
The next line calls F.getName.  Is your concern something like C++ name mangling?

arsenm replied:
My concern is anonymous functions, and function names changed by optimizations. You can unconditionally use printAsOperand without pre-checking the name is empty

jdenny-ornl replied:
> My concern is anonymous functions, and function names changed by optimizations.

I've changed it to use debug info if available.

> You can unconditionally use printAsOperand without pre-checking the name is empty

If debug info is not available, it now uses the LLVM name, including the `@`, as printed by printAsOperand.  It still checks whether `Name` is empty because that means the callee either wasn't a function or didn't have the desired metadata.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,345 @@
+        if (SubProgram->isArtificial())
+          R << " artificial";
+      }
+      if (!Name.empty())
+        R << " '" << Name << "'";
+      else
+        R << " with unknown name";
```
If we have to, better to print the anonymous function reference. You can maybe use printAsOperand for this?

jdenny-ornl replied:
Done.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,345 @@
+    } else {
+      R << ", '" << Inst.getOpcodeName() << "' instruction";
+    }
+    if (Inst.hasName())
+      R << " ('%" << Inst.getName() << "')";
```
printAsOperand?

jdenny-ornl replied:
Done.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,345 @@
+        CallKind += " call";
+        RemarkKind += "Call";
+      }
+      if (!Call->isIndirectCall()) {
+        if (const Function *Callee = Call->getCalledFunction()) {
```
Using both isIndirectCall and null checking getCalledFunction is confusing, and I'm not sure how these are supposed to interact. They're almost redundant except for calls to direct constants and inline asm, which probably should be handled explicitly?

jdoerfert replied:
We likely want to flag asm independently. Later we could try to match it to builtins where we have them to help the compiler (and the user).

jdenny-ornl replied:
> Using both isIndirectCall and null checking getCalledFunction is confusing, and I'm not sure how these are supposed to interact. 

I believe I saw the getCalledFunction comments and wasn't sure how to handle the signature issue.  So I made sure I wouldn't end up dereferencing a null pointer:

```
Returns the function called, or null if this is an indirect function invocation or the function signature does not match the call signature.
```

> We likely want to flag asm independently.

Done.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,345 @@
+      }
+      if (!Call->isIndirectCall()) {
+        if (const Function *Callee = Call->getCalledFunction()) {
+          if (Callee && !Callee->isIntrinsic() && !Callee->isDeclaration()) {
```
Double null check of Callee

jdenny-ornl replied:
Done.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,345 @@
+        FlatAddrspaceAccesses += Direction;
+        remarkFlatAddrspaceAccess(ORE, F, I);
+      }
+    }
```
You should also consider getTgtMemIntrinsic and/or TTI::collectFlatAddressOperands

jdenny-ornl replied:
What's an example of a GPU intrinsic that can use the flat address space?

arsenm replied:
Well the degenerate cases are the llvm.amdgcn.is.shared/llvm.amdgcn.is.private intrinsics, which aren't real memory accesses. And the other cases we handle I have outstanding patches to delete, so I guess this doesn't matter much

jdenny-ornl replied:
I did write a patch using getTgtMemIntrinsic, and I can add it if you think cases might appear in the future.  To test it, I temporarily modified AMDGPUDSOrderedIntrinsic to accept address space 0.  I obviously should not upstream that part, and I didn't find any other way to test it.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,345 @@
+}
+
+static std::vector<std::optional<int64_t>>
+parseFnAttrAsIntegerFields(Function &F, StringRef Name, unsigned NumFields) {
```
Probably should move this into Function along with getFnAttributeAsParsedInteger

jdenny-ornl replied:
I look into code reorg suggestions like this after the issue on effective vs. explicit attributes is settled.

jdenny-ornl replied:
The only thing left here is parseFnAttrAsInteger, which is minimal.  Would you like to see that moved  into Function?

jdenny-ornl committed:
Include less in header

jdenny-ornl committed:
Removed unused comparison operators

They wouldn't have worked reliably anyway given uninitialized padding
in the struct.

jdenny-ornl committed:
Remove redundant null check

jdoerfert commented:
> Short-term you could add an Offload.rst and link everything from there. Long-term you may ask for offload.llvm.org.

Yes, offload.llvm.org is on my list, I started to talk to people and dropped the ball.
The Readme we have is a place we can put some information for now, and openmp.llvm.org as well. Eventually that will be split off.

tschuett commented:
> > Short-term you could add an Offload.rst and link everything from there. Long-term you may ask for offload.llvm.org.
> 
> Yes, offload.llvm.org is on my list, I started to talk to people and dropped the ball. The Readme we have is a place we can put some information for now, and openmp.llvm.org as well. Eventually that will be split off.

I was talking about an Offload.rst in https://github.com/llvm/llvm-project/tree/main/llvm/docs for better visibility.

jdenny-ornl committed:
Move KernelInfo to KernelInfo.cpp, remove KernelInfoAnalysis

For now, analysis results will not be used beyond emitting remarks.
If that changes, we can revert.

jdenny-ornl committed:
Use printAsOperand not getName to identify instruction

jdenny-ornl committed:
Use printAsOperand to report indirect callee

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -84,20 +147,21 @@ static void remarkCall(OptimizationRemarkEmitter &ORE, const Function &Caller,
-      if (auto *SubProgram = Callee->getSubprogram()) {
+    R << ", " << CallKind << ", callee is";
+    Value *Callee = Call.getCalledOperand();
+    std::string Name;
```
SmallString?

jdenny-ornl replied:
Done.

jdenny-ornl committed:
Report inline assembly calls

jdenny-ornl committed:
Use llvm::SmallString

jdenny-ornl commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,425 @@
+      R << ", '" << Inst.getOpcodeName() << "' instruction";
+    }
+    if (!Inst.getType()->isVoidTy()) {
+      std::string Name;
```
Oops missed one.

jdenny-ornl replied:
Done.  Feel free to suggest different lengths for this small strings.

jdenny-ornl committed:
Use llvm::SmallString

jdenny-ornl committed:
getKernelInfo -> emitKernelInfo because return is unused

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl commented:
> > > Short-term you could add an Offload.rst and link everything from there. Long-term you may ask for offload.llvm.org.
> > 
> > 
> > Yes, offload.llvm.org is on my list, I started to talk to people and dropped the ball. The Readme we have is a place we can put some information for now, and openmp.llvm.org as well. Eventually that will be split off.
> 
> I was talking about an Offload.rst in https://github.com/llvm/llvm-project/tree/main/llvm/docs for better visibility.

Do either of you want to see any additional documentation change in this PR?

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl committed:
Clean up launch bounds

* For amdgpu, use AMGPUSubtarget functions to query values.  Thus, we
  end up with logical values that don't appear explicitly in the IR,
  and we ignore some impossible values that do appear explicitly.
* For nvptx, use NVPTXUtilities.h functions to query values.  Thus,
  drop KernelInfo.cpp's implementation of NVVM annotation parsing.
  Also, add support for a few more launch bounds.
* Move target-specific collection of launch bounds to target-specific
  classes (GCNSubtarget and NVPTXSubtarget).  While making the above
  changes, I struggled to find the right headers to enable keeping the
  implementation in KernelInfo.cpp, and one reviewer wanted to see
  some reorganization along these lines anyway.

arsenm commented on llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:
```diff
@@ -338,6 +338,11 @@ class TargetSubtargetInfo : public MCSubtargetInfo {
   virtual bool supportsInitUndef() const { return false; }
+
+  /// For \p F, call \p Body with the name and value of each launch bound.
+  virtual void forEachLaunchBound(
+      const Function &F,
+      std::function<void(StringRef Name, unsigned Value)> Body) const {}
```
It would be more appropriate to add something to TargetTransformInfo. It also needs a more specific name and documentation comment. Also should use int64_t for arbitrary integer values 

This can also be function_ref

jdenny-ornl replied:
> It would be more appropriate to add something to TargetTransformInfo.

Done.

> It also needs a more specific name and documentation comment.

I'm fine to change it, but what details do you want the name and comment to reveal?

> Also should use int64_t for arbitrary integer values
> 
> This can also be function_ref

Done.

jdenny-ornl replied:
Everything appears to be addressed except [the part that continues here](https://github.com/llvm/llvm-project/pull/102944#discussion_r1773738493).  Resolving.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -298,90 +284,30 @@ static void remarkProperty(OptimizationRemarkEmitter &ORE, const Function &F,
-  if (MDNode *ExistingOp = getNVPTXMDNode(F, Name)) {
-    auto *Op = cast<ConstantAsMetadata>(ExistingOp->getOperand(2));
-    Result = cast<ConstantInt>(Op->getValue())->getZExtValue();
+  Attribute A = F.getFnAttribute(Name);
+  if (!A.isStringAttribute())
+    return std::nullopt;
+  StringRef Field = A.getValueAsString();
+  int64_t Val;
+  if (Field.getAsInteger(0, Val)) {
+    F.getContext().emitError("cannot parse integer in attribute '" + Name +
+                             "': " + Field);
+    return std::nullopt;
```
This is reinventing F.getFnAttributeAsParsedInteger

jdenny-ornl replied:
I rewrote it to call getFnAttributeAsParsedInteger.

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl committed:
Adjust forEachLaunchBound param

* std::function -> llvm::function_ref
* unsigned -> int64_t

jdenny-ornl committed:
Reuse Function::getFnAttributeAsParsedInteger

jdenny-ornl committed:
Move forEachLaunchBound to TargetTransformInfo

arsenm commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1788,6 +1788,11 @@ class TargetTransformInfo {
 
   /// @}
 
+  /// For \p F, call \p Body with the name and value of each launch bound.
+  void forEachLaunchBound(
```
This is a very specific, not general concept. The name and comment needs to elaborate for what the context is.

Could this be a query to return the set of attributes and values, rather than the callback?

jdenny-ornl replied:
> This is a very specific, not general concept. The name and comment needs to elaborate for what the context is.

@jdoerfert Do you have any suggestion about the right way to document this?  Are there any additional attributes you would like for GCNTTIImpl::collectLaunchBounds or NVPTXTTIImpl::collectLaunchBounds to collect?

> Could this be a query to return the set of attributes and values, rather than the callback?

Done.

jdenny-ornl replied:
The naming discussion continues at: https://github.com/llvm/llvm-project/pull/102944#discussion_r1795842757

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

Update NVPTXTargetTransformInfo.cpp's getMaxClusterRank call for
interface changes made by 489acb2401b5.

jdenny-ornl committed:
forEachLaunchBound -> collectLaunchBounds

Return the launch bounds instead of passing them to a callback.

tschuett commented on llvm/include/llvm/Analysis/KernelInfo.h:
```diff
@@ -0,0 +1,36 @@
+class TargetMachine;
+
+class KernelInfoPrinter : public PassInfoMixin<KernelInfoPrinter> {
+private:
```
The `private:` has no effect.

jdenny-ornl replied:
Removed.

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl committed:
Remove redundant private

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1390,3 +1390,19 @@ unsigned GCNTTIImpl::getPrefetchDistance() const {
+  auto AmdgpuWavesPerEU = ST->getWavesPerEU(F);
+  LB.push_back({"AmdgpuWavesPerEUMin", AmdgpuWavesPerEU.first});
+  LB.push_back({"AmdgpuWavesPerEUMax", AmdgpuWavesPerEU.second});
+  // TODO: Any others we should add?
```
Remove the todo

jdenny-ornl replied:
Done.

arsenm commented on llvm/lib/Analysis/TargetTransformInfo.cpp:
```diff
@@ -1354,6 +1354,12 @@ bool TargetTransformInfo::hasActiveVectorLength(unsigned Opcode, Type *DataType,
   return TTIImpl->hasActiveVectorLength(Opcode, DataType, Alignment);
 }
 
+void TargetTransformInfo::collectLaunchBounds(
```
Maybe collectKernelLaunchBounds?

jdenny-ornl replied:
Done.

arsenm commented on llvm/lib/Target/TargetMachine.cpp:
```diff
@@ -26,6 +26,11 @@
 #include "llvm/Target/TargetLoweringObjectFile.h"
 using namespace llvm;
 
+cl::opt<bool> KernelInfoEndLTO(
```
Can you just unconditionally run the pass, and early exit if !isAnalysisRemarkEnabled?

jdenny-ornl replied:
Done.  But I converted -kernel-info-end-lto to -no-kernel-info-end-lto so it's still possible to experiment with positioning kernel-info elsewhere in a pass pipeline, as described in KernelInfo.rst.

jdenny-ornl replied:
Discussion continued [later](https://github.com/llvm/llvm-project/pull/102944#discussion_r1799726914).  Resolving this thread.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -781,6 +782,15 @@ void AMDGPUTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
+    PB.registerFullLinkTimeOptimizationLastEPCallback(
+        [this](ModulePassManager &PM, OptimizationLevel Level) {
+          FunctionPassManager FPM;
+          FPM.addPass(KernelInfoPrinter(this));
```
There's already a registerFullLinkTimeOptimizationLastEPCallback call above, move this into it

jdenny-ornl replied:
Done.

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl committed:
Remove todos, as requested

jdenny-ornl committed:
Combine registerFullLinkTimeOptimizationLastEPCallback calls

jdenny-ornl committed:
collectLaunchBounds -> collectKernelLaunchBounds

jdenny-ornl committed:
Spell kernel-info properties like their IR attributes

jdenny-ornl committed:
Replace -kernel-info-end-lto with -no-kernel-info-end-lto

jdenny-ornl commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1390,3 +1390,19 @@ unsigned GCNTTIImpl::getPrefetchDistance() const {
+  auto AmdgpuMaxNumWorkgroups = ST->getMaxNumWorkGroups(F);
+  LB.push_back({"AmdgpuMaxNumWorkgroupsX", AmdgpuMaxNumWorkgroups[0]});
+  LB.push_back({"AmdgpuMaxNumWorkgroupsY", AmdgpuMaxNumWorkgroups[1]});
+  LB.push_back({"AmdgpuMaxNumWorkgroupsZ", AmdgpuMaxNumWorkgroups[2]});
```
Originally it was to have consistency among property names in kernel-info remarks. But that doesn't seem too important, especially now that this part of the code is outside kernel-info.  Done.

jdenny-ornl committed:
Apply clang-format

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -824,6 +825,11 @@ void AMDGPUTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
           PM.addPass(AMDGPULowerModuleLDSPass(*this));
         if (EnableAMDGPUAttributor && Level != OptimizationLevel::O0)
           PM.addPass(AMDGPUAttributorPass(*this));
+        if (!NoKernelInfoEndLTO) {
```
I don't see the point of the option

jdenny-ornl replied:
See [this comment](https://github.com/llvm/llvm-project/pull/102944#discussion_r1797494663) and [these docs](https://github.com/llvm/llvm-project/blob/d3beccfe9eb34636cd0015b867d6fcda8fa6ea26/llvm/docs/KernelInfo.rst?plain=1#L43-L63).

jdenny-ornl replied:
Resolving this thread under the assumption my clarification was sufficient.  Please reopen if necessary.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:
```diff
@@ -1410,3 +1410,18 @@ unsigned GCNTTIImpl::getPrefetchDistance() const {
+void GCNTTIImpl::collectKernelLaunchBounds(
+    const Function &F,
+    SmallVectorImpl<std::pair<StringRef, int64_t>> &LB) const {
+  auto MaxNumWorkgroups = ST->getMaxNumWorkGroups(F);
```
no auto

jdenny-ornl replied:
Done.

arsenm commented on llvm/test/Analysis/KernelInfo/allocas.ll:
```diff
@@ -0,0 +1,78 @@
+
+define void @h() !dbg !3 {
+entry:
+  ; CHECK: remark: test.c:0:0: in artificial function 'h', artificial alloca 'dyn_ptr' with static size of 8 bytes
```
Not sure what's artificial about it

jdenny-ornl replied:
This test makes sure that functions marked as artificial in metadata are reported that way.  It's been a while, but my recollection is I started with IR generated by clang, ran it through llvm-reduce, and then further reduced it by hand.

What are you objecting to?  Do you want the test function to look more like an artificial function in some way?  Do you want artificial functions not to be called out by kernel-info?

arsenm commented on llvm/test/Analysis/KernelInfo/allocas.ll:
```diff
@@ -0,0 +1,78 @@
+  %i = alloca i32, align 4
+  ; CHECK: remark: test.c:15:9: in artificial function 'h', alloca 'a' with static size of 8 bytes
+  %a = alloca [2 x i32], align 4
+  tail call void @llvm.dbg.declare(metadata ptr %dyn_ptr.addr, metadata !7, metadata !DIExpression()), !dbg !11
```
Test variable alloca? Also  consider alloca outside of the entry block as dynamic

jdenny-ornl replied:
> Test variable alloca?

Sorry, I don't understand.  What is your suggestion?

> Also consider alloca outside of the entry block as dynamic

You want me to test an alloca that is not in the entry block and that does not have a static size?

jdenny-ornl replied:
Done.

jdenny-ornl committed:
Avoid auto, as requested

jdenny-ornl committed:
For function name, use debug info or keep @

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,330 @@
+#define DEBUG_TYPE "kernel-info"
+
+/// Data structure holding function info for kernels.
+class KernelInfo {
```
Should be in anonymous namespace

jdenny-ornl replied:
Done.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,330 @@
+    raw_svector_ostream OS(Name);
+    V->printAsOperand(OS, /*PrintType=*/false, M);
+  }
+  if (!Kind.empty())
+    R << Kind << " ";
+  R << "'" << Name << "'";
```
```suggestion
  if (!Kind.empty())
    R << Kind << ' ';
  R << '\'' << Name << '\'';
```

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,330 @@
+  const DominatorTree &DT = FAM.getResult<DominatorTreeAnalysis>(F);
+  auto &ORE = FAM.getResult<OptimizationRemarkEmitterAnalysis>(F);
+  for (const auto &BB : F)
+    if (DT.isReachableFromEntry(&BB))
```
Why does this need to worry about unreachable blocks? This isn't doing any transforms that need to consider breaking the IR

jdenny-ornl replied:
I copied that from FunctionPropertiesAnalysis.cpp and assumed it might eventually become a use case.  Removed for now.

arsenm commented on llvm/lib/Analysis/KernelInfo.cpp:
```diff
@@ -0,0 +1,330 @@
+
+void KernelInfo::updateForBB(const BasicBlock &BB, int64_t Direction,
+                             OptimizationRemarkEmitter &ORE) {
+  assert(Direction == 1 || Direction == -1);
```
What's the point of this Direction operand? These values will never decrease?

jdenny-ornl replied:
I copied that from FunctionPropertiesAnalysis.cpp and assumed it might eventually become a use case.  Removed for now.

arsenm commented on llvm/test/Analysis/KernelInfo/enable-kernel-info/amdgpu.ll:
```diff
@@ -0,0 +1,18 @@
+; DEFINE: %{fcheck-on} = FileCheck -match-full-lines %S/Inputs/test.ll
+; DEFINE: %{fcheck-off} = FileCheck -allow-empty -check-prefixes=NONE \
+; DEFINE:                 %S/Inputs/test.ll
+
```
Test the bounds attributes?

jdenny-ornl replied:
I don't know what you're asking.  This test checks when kernel-info is enabled by looking for its output, and it arbitrarily chooses `omp_target_num_teams` as the output to look for.  Do you want more bounds checking?  Do you want something else intsead?

arsenm commented on llvm/test/Analysis/KernelInfo/enable-kernel-info/amdgpu.ll:
```diff
@@ -0,0 +1,18 @@
+; Check when kernel-info is enabled in the AMD GPU target backend.
+
```
As it is now, this is not a .ll test

jdenny-ornl replied:
Renamed to .test.

jdenny-ornl committed:
Use anonymous namespace

jdenny-ornl committed:
Remove currently unused capabilities, as requested

They were originally copied from FunctionPropertiesAnalysis.cpp.

jdenny-ornl committed:
Rename test files without LLVM IR to .test

jdenny-ornl committed:
Regenerate OpenMP tests from current clang

See llvm/test/Analysis/KernelInfo/openmp/README.md.

jdenny-ornl committed:
Include LLVM value name in alloca report

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl committed:
Update expected amdgpu-max-num-workgroups default values

Due to 0b40f979298a.

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl committed:
Regenerate OpenMP tests from current clang

See llvm/test/Analysis/KernelInfo/openmp/README.md.

jdenny-ornl commented:
> The following files introduce new uses of undef:
> 
>     * llvm/test/Analysis/KernelInfo/openmp/nvptx.ll

We clang stops generating those uses, we can update the tests.

jdoerfert reviewed (commented):
Two minor issues and then we can merge this.

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl committed:
Relocate and use llvm::omp::getDeviceKernels

jdenny-ornl committed:
Extend test to cover dyn and non-entry allocas

AlexMaclean commented on llvm/lib/IR/Module.cpp:
```diff
@@ -322,6 +322,32 @@ void Module::eraseNamedMetadata(NamedMDNode *NMD) {
 
+SetVector<Function *> Module::getDeviceKernels() {
+  // TODO: Create a more cross-platform way of determining device kernels.
+  NamedMDNode *MD = getNamedMetadata("nvvm.annotations");
```
I'm working on deprecating this method of specifying a kernel in favor of the `ptx_kernel` calling convention (see https://github.com/llvm/llvm-project/pull/120806). I think it would be good to check the calling convention of all functions in the module and add them to the set if they have `ptx_kernel` or `amdgpu_kernel` (maybe others too?).

jdenny-ornl replied:
Thanks for letting me know.  Now that your PR #122320 has landed, I've updated this PR to use it.

I see that `llvm::omp::getDeviceKernels` checks for the kernel calling convention but also still kernel in nvvm.annotations.  Should KernelInfo check for the latter as well?  My update only checks for the former.

AlexMaclean replied:
https://github.com/llvm/llvm-project/pull/119261 will auto-upgrade nvvm.annotations allowing you to only check the CC. Currently you'll still need to check both but hopefully soon only CC will be needed.

jdenny-ornl replied:
When is that needed?  Just for previously generated LLVM IR?  Clang at today's main seems to generate the calling conventions already.

AlexMaclean replied:
Yea to maintain backwards compatibility with older IR, or out-of-tree front-ends.

jdenny-ornl replied:
Makes sense.  But for KernelInfo, I'm not sure if it's worth it.  We could probably live with that small shortcoming until PR #119261 lands.  @jdoerfert?

jdoerfert replied:
Backwards compatibility with old GPU IR is not much of a thing. This should be fine.

jdenny-ornl committed:
Revert "Relocate and use llvm::omp::getDeviceKernels"

This reverts commit bb9d5c20c869d3429ee8c9ea3c16ca2551bb2734.

This will facilitate merging main due to 07ed818 (PR #122320), which
changes llvm::omp::getDeviceKernels.  Will rewrite and reapply after
merging main.

jdenny-ornl committed:
Merge branch 'main' into kernel-info-pr

jdenny-ornl committed:
Relocate and use OpenMPOpt.cpp's isKernelCC

Also, regenerate OpenMP tests from current clang so they see the new
kernel calling conventions.

jdoerfert reviewed (approved):
LG, I think.

slackito commented:
I was trying to fix the bazel build after this change, and it seems it introduces a couple of circular dependencies. `Analysis/KernelInfo.cpp` includes `Passes/PassBuilder.h` and `Target/TargetMachine.h`, but both `Passes` and `Target` already depend on `Analysis`.

I think the circular dependency can be broken by removing the PassBuilder.h include, which seems to be unused, and replacing the TargetMachine include with a forward decl. Something like this:
```
diff --git a/llvm/lib/Analysis/KernelInfo.cpp b/llvm/lib/Analysis/KernelInfo.cpp
index be9b1136321f..5175614645bf 100644
--- a/llvm/lib/Analysis/KernelInfo.cpp
+++ b/llvm/lib/Analysis/KernelInfo.cpp
@@ -15,17 +15,20 @@
 #include "llvm/ADT/SmallString.h"
 #include "llvm/ADT/StringExtras.h"
 #include "llvm/Analysis/OptimizationRemarkEmitter.h"
+#include "llvm/Analysis/TargetTransformInfo.h"
 #include "llvm/IR/DebugInfo.h"
 #include "llvm/IR/Dominators.h"
 #include "llvm/IR/Instructions.h"
 #include "llvm/IR/Metadata.h"
 #include "llvm/IR/Module.h"
 #include "llvm/IR/PassManager.h"
-#include "llvm/Passes/PassBuilder.h"
-#include "llvm/Target/TargetMachine.h"
 
 using namespace llvm;
 
+namespace llvm {
+class TargetMachine;
+}
+
 #define DEBUG_TYPE "kernel-info"
 
 namespace {
```
